### ISSUE TITLE
Possible shipping amount rounding problem from WC 2.3.8 update..??

### ISSUE THREAD

**Issue by angelleye** [angelleye/paypal-woocommerce#203]
https://wordpress.org/support/topic/10427-shipping-total-is-invalid-1?replies=3#post-6871878

I've heard this same thing from another person recently, too.

**Comment by angelleye** [comment]
The user in that thread says he fixed it by simply rounding the shipping value in the wc-gateway-paypal-pro-angelleye.php class.  

He says the problem isn't happening with Express Checkout, so I assume we were doing the rounding there but not in Pro..??

Need to look into that and make sure we're rounding in case the value passed from WC isn't rounded for us.  

Also need to check the PayFlow class and see if the same thing needs to be done there or not..??

**Comment by zefrenchemen** [comment]
I've updated the thread to reflect that this was a premature statement :)
While we did fix what we perceived to be the problem - the Canada Post passing along 3 decimals - we ended up being met with what was in fact the core of the problem : since WC 2.3.8, rounding of certain values is broken.
In our case it's broken in two ways : while fetching the shipping charges and passing them to PP, it doesn't round to 2 decimals. If forced to do so, then it improperly rounds up instead of half-up when calculating taxes. So every x transaction, one comes up to be 1 cent higher than PP's expected total. This last issue does not always occur. For it to happen, a number (shipping, taxes, total, etc) must have been rounded up before half, that is .1, .2, .3 and .4 where all rounded up when they should have been rounded down.

After flipping this every which way, it does look like the problem is coming from PP for WC plugin, but the cause seems to be WC changing how they deal with rounding in this last update. As you state above, all values are passed properly in PP standard. This problem occurs only with PP Pro, unfortunately for us, that's 98% of business :(  We've temporarily fixed this by enabling fixed-rate shipping, a costly workaround.
